File: sdk/python-sdk/verity_sdk/protocols/v1_0/Relationship.py
from enum import Enum

from verity_sdk.protocols.Protocol import Protocol
from verity_sdk.utils import EVERNYM_MSG_QUALIFIER, Context


class GoalsList(Enum):
    """An enumeration of possible goals (reasons) for a relationship."""

    class GoalCode:
        """
        a tuple with a short code string and human readable name string
        """

        def __init__(self, code, name) -> None:
            """
            Args:
                code (str): short code string
                name (str): human readable string
            """
            self.code = code
            self.name = name

    ISSUE_VC = GoalCode('issue-vc', 'To issue a credential')
    """To issue a credential"""

    REQUEST_PROOF = GoalCode('request-proof', 'To request a proof')
    """To request a proof"""

    CREATE_ACCOUNT = GoalCode('create-account', 'To create an account with a service')
    """To create an account with a service"""

    P2P_MESSAGING = GoalCode('p2p-messaging', 'To establish a peer-to-peer messaging relationship')
    """To establish a peer-to-peer messaging relationship"""


class Relationship(Protocol):
    """
    The Relationship protocol creates and manages relationships on the verity-application agent. These relationships
    are secure communication channels between self-sovereign parties. A relationship much be created before using the
    connections protocols. In the future this protocol will allow management of each given relationship
    (eg key rotation)
    """
    MSG_FAMILY = 'relationship'
    """the family name for the message family"""

    MSG_FAMILY_VERSION = '1.0'
    """the version for the message family"""

    CREATED = 'created'
    """Name for 'created' signal message"""
    INVITATION = 'invitation'
    """Name for 'invitation' signal message"""

    CREATE = 'create'
    """Name for 'create' control message"""
    CONNECTION_INVITATION = 'connection-invitation'
    """Name for 'connection-invitation' control message"""
    OUT_OF_BAND_INVITATION = 'out-of-band-invitation'
    """Name for 'out-of-band-invitation' control message"""
    SMS_CONNECTION_INVITATION = 'sms-connection-invitation'
    """Name for 'sms-connection-invitation' control message"""
    SMS_OUT_OF_BAND_INVITATION = 'sms-out-of-band-invitation'
    """Name for 'sms-out-of-band-invitation' control message"""

    def __init__(self,
                 for_relationship: str = None,
                 thread_id: str = None,
                 label: str = None,
                 logo_url: str = None,
                 phone_number: str = None
                 ):
        """
        Args:
            for_relationship (str): the relationship identifier (DID) for the pairwise relationship that will be used
            thread_id (str): the thread id of the already started protocol
            label (str): he label presented in the invitation to connect to this relationship
            logo_url (str): logo url presented in invitation
            phone_number (str): Mobile phone number in international format, eg. +18011234567
        """
        super().__init__(
            self.MSG_FAMILY,
            self.MSG_FAMILY_VERSION,
            msg_qualifier=EVERNYM_MSG_QUALIFIER,
            thread_id=thread_id
        )

        self.for_relationship = for_relationship
        self.label = label
        self.logo_url = logo_url
        self.phone_number = phone_number

    async def create(self, context):
        """
        Directs verity-application to create a new relationship

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.create_msg_packed(context))

    def create_msg(self, context: Context):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.CREATE)
        self._add_thread(msg)
        if self.label:
            msg['label'] = self.label
        if self.logo_url:
            msg['logoUrl'] = self.logo_url
        if self.phone_number:
            msg['phoneNumber'] = self.phone_number

        return msg

    async def create_msg_packed(self, context: Context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.create_msg(context))

    async def connection_invitation(self, context: Context, short_invite: bool = None):
        """
        Ask for aries invitation from the verity-application agent for the relationship created by this protocol

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            short_invite (bool): Request the short invite
        """
        await self.send_message(context, await self.connection_invitation_msg_packed(context, short_invite))

    def connection_invitation_msg(self, context: Context, short_invite: bool = None):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            short_invite (bool): Request the short invite
        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.CONNECTION_INVITATION)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        if short_invite is not None:
            msg['shortInvite'] = short_invite
        return msg

    async def connection_invitation_msg_packed(self, context: Context, short_invite: bool = None):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            short_invite (bool): Request the short invite
        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.connection_invitation_msg(context, short_invite))

    async def sms_connection_invitation(self, context: Context):
        """
        Ask for SMS aries invitation from the verity-application agent for the relationship created by this protocol

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.sms_connection_invitation_msg_packed(context))

    def sms_connection_invitation_msg(self, context: Context):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.SMS_CONNECTION_INVITATION)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        return msg

    async def sms_connection_invitation_msg_packed(self, context: Context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.sms_connection_invitation_msg(context))

    async def out_of_band_invitation(self,
                                     context: Context,
                                     short_invite: bool = None,
                                     goal: GoalsList = GoalsList.P2P_MESSAGING):
        """
        Ask for aries out of band invitation from the verity-application agent for the relationship
        created by this protocol

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            short_invite (bool): Request the short invite
            goal (GoalsList): the initial intended goal of the relationship (this goal is expressed in the invite)
        """
        await self.send_message(context, await self.out_of_band_invitation_msg_packed(context, short_invite, goal))

    def out_of_band_invitation_msg(self,
                                   context: Context,
                                   short_invite: bool = None,
                                   goal: GoalsList = GoalsList.P2P_MESSAGING):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            short_invite (bool): Request the short invite
            goal (GoalsList): the initial intended goal of the relationship (this goal is expressed in the invite)
        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.OUT_OF_BAND_INVITATION)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        if short_invite is not None:
            msg['shortInvite'] = short_invite

        if goal:
            msg['goalCode'] = goal.value.code
            msg['goal'] = goal.value.name

        return msg

    async def out_of_band_invitation_msg_packed(self,
                                                context: Context,
                                                short_invite: bool = None,
                                                goal: GoalsList = GoalsList.P2P_MESSAGING):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            short_invite (bool): Request the short invite
            goal (GoalsList): the initial intended goal of the relationship (this goal is expressed in the invite)
        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.out_of_band_invitation_msg(context, short_invite, goal))

    async def sms_out_of_band_invitation(self,
                                         context: Context,
                                         goal: GoalsList = GoalsList.P2P_MESSAGING):
        """
        Ask for SMS aries out of band invitation from the verity-application agent for the relationship
        created by this protocol

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            goal (GoalsList): the initial intended goal of the relationship (this goal is expressed in the invite)
        """
        await self.send_message(context, await self.sms_out_of_band_invitation_msg_packed(context, goal))

    def sms_out_of_band_invitation_msg(self,
                                       context: Context,
                                       goal: GoalsList = GoalsList.P2P_MESSAGING):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            goal (GoalsList): the initial intended goal of the relationship (this goal is expressed in the invite)
        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.SMS_OUT_OF_BAND_INVITATION)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)

        if goal:
            msg['goalCode'] = goal.value.code
            msg['goal'] = goal.value.name

        return msg

    async def sms_out_of_band_invitation_msg_packed(self,
                                                    context: Context,
                                                    goal: GoalsList = GoalsList.P2P_MESSAGING):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            goal (GoalsList): the initial intended goal of the relationship (this goal is expressed in the invite)
        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.sms_out_of_band_invitation_msg(context, goal))

File: sdk/nodejs-sdk/src/protocols/v1_0/Relationship.js
'use strict'
const utils = require('../../utils')
const Protocol = require('../Protocol')

/**
 * An interface for controlling a 1.0 Relationship protocol.
 * @extends Protocol
 */
class Relationship extends Protocol {
  /**
   * Constructor for the 1.0 Relationship object. This constructor creates an object that is ready to create a new
   * relationship.
   * @param forRelationship the relationship identifier (DID) for the pairwise relationship that will be used
   * @param threadId the thread id of the already started protocol
   * @param label the label presented in the invitation to connect to this relationship
   * @param logoUrl logo url presented in invitation
   * @return 1.0 Relationship object
   *
   * @property {String} msgFamily - 'relationship'
   * @property {String} msgFamilyVersion - '1.0'
   * @property {String} msgQualifier - 'Community Qualifier'
   * @property {String} this.msgNames.CREATE - 'create'
   * @property {String} this.msgNames.CONNECTION_INVITATION - 'connection-invitation'
   * @property {String} this.msgNames.OUT_OF_BAND_INVITATION - 'out-of-band-invitation'
   * @property {String} this.msgNames.SMS_CONNECTION_INVITATION - 'sms-connection-invitation'
   * @property {String} this.msgNames.SMS_OUT_OF_BAND_INVITATION - 'sms-out-of-band-invitation'
   * @property {String} this.msgNames.CREATED - 'created'
   * @property {String} this.msgNames.INVITATION - 'invitation'
   */
  constructor (forRelationship = null, threadId = null, label = null, logoUrl = null) {
    const msgFamily = 'relationship'
    const msgFamilyVersion = '1.0'
    const msgQualifier = utils.constants.EVERNYM_MSG_QUALIFIER
    super(msgFamily, msgFamilyVersion, msgQualifier, threadId)

    this.msgNames.CREATE = 'create'
    this.msgNames.CONNECTION_INVITATION = 'connection-invitation'
    this.msgNames.OUT_OF_BAND_INVITATION = 'out-of-band-invitation'
    this.msgNames.SMS_CONNECTION_INVITATION = 'sms-connection-invitation'
    this.msgNames.SMS_OUT_OF_BAND_INVITATION = 'sms-out-of-band-invitation'
    this.msgNames.CREATED = 'created'
    this.msgNames.INVITATION = 'invitation'

    this.forRelationship = forRelationship
    this.label = label
    this.logoUrl = logoUrl
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #create
     */
  async createMsg (context) {
    var msg = this._getBaseMessage(this.msgNames.CREATE)
    msg = this._addThread(msg)
    if (this.label) {
      msg.label = this.label
    }
    if (this.logoUrl) {
      msg.logoUrl = this.logoUrl
    }
    return msg
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #create
     */
  async createMsgPacked (context) {
    return this.getMessageBytes(context, await this.createMsg(context))
  }

  /**
     * Directs verity-application to create a new relationship
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async create (context) {
    await this.sendMessage(context, await this.createMsgPacked(context))
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @param shortInvite enables short version of invite to be generated
     * @return the constructed message (JSON object)
     *
     * @see #connectionInvitation
     */
  async connectionInvitationMsg (context, shortInvite = null) {
    var msg = this._getBaseMessage(this.msgNames.CONNECTION_INVITATION)
    msg['~for_relationship'] = this.forRelationship
    msg = this._addThread(msg)
    if (shortInvite != null) {
      msg.shortInvite = shortInvite
    }
    return msg
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @param shortInvite enables short version of invite to be generated
     * @return the byte array ready for transport
     *
     * @see #connectionInvitation
     */
  async connectionInvitationMsgPacked (context, shortInvite = null) {
    return this.getMessageBytes(context, await this.connectionInvitationMsg(context, shortInvite))
  }

  /**
     * Ask for aries invitation from the verity-application agent for the relationship created by this protocol
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @param shortInvite enables short version of invite to be generated
     */
  async connectionInvitation (context, shortInvite = null) {
    await this.sendMessage(context, await this.connectionInvitationMsgPacked(context, shortInvite))
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @param phoneNumber mobile phone number in international format (eg. +18011234567) which can be used for sending SMS invitations.
     * @return the constructed message (JSON object)
     *
     * @see #smsConnectionInvitation
     */
  async smsConnectionInvitationMsg (context, phoneNumber) {
    var msg = this._getBaseMessage(this.msgNames.SMS_CONNECTION_INVITATION)
    msg['~for_relationship'] = this.forRelationship
    msg.phoneNumber = phoneNumber
    msg = this._addThread(msg)
    return msg
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @param phoneNumber mobile phone number in international format (eg. +18011234567) which can be used for sending SMS invitations.
     * @return the byte array ready for transport
     *
     * @see #smsConnectionInvitation
     */
  async smsConnectionInvitationMsgPacked (context, phoneNumber) {
    return this.getMessageBytes(context, await this.smsConnectionInvitationMsg(context, phoneNumber))
  }

  /**
     * Ask for aries invitation from the verity-application agent for the relationship created by this protocol
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @param phoneNumber mobile phone number in international format (eg. +18011234567) which can be used for sending SMS invitations.
     */
  async smsConnectionInvitation (context, phoneNumber) {
    await this.sendMessage(context, await this.smsConnectionInvitationMsgPacked(context, phoneNumber))
  }

  /**
   * Creates the control message without packaging and sending it.
   * @param context an instance of the Context object initialized to a verity-application agent
   * @param shortInvite enables short version of invite to be generated
   * @param goal goal code for providing the goal of the connection
   * @return the constructed message (JSON object)
   *
   * @see #outOfBandInvitation
   */
  async outOfBandInvitationMsg (context, shortInvite = null, goal = null) {
    var msg = this._getBaseMessage(this.msgNames.OUT_OF_BAND_INVITATION)
    msg['~for_relationship'] = this.forRelationship
    msg = this._addThread(msg)
    if (shortInvite != null) {
      msg.shortInvite = shortInvite
    }

    if (goal != null) {
      msg.goalCode = goal.code
      msg.goal = goal.goalName
    }
    return msg
  }

  /**
   * Creates and packages message without sending it.
   * @param context an instance of the Context object initialized to a verity-application agent
   * @param shortInvite enables short version of invite to be generated
   * @param goal goal code for providing the goal of the connection
   * @return the byte array ready for transport
   *
   * @see #outOfBandInvitation
   */
  async outOfBandInvitationMsgPacked (context, shortInvite = null, goal = null) {
    return this.getMessageBytes(context, await this.outOfBandInvitationMsg(context, shortInvite, goal))
  }

  /**
   * Ask for aries out of band invitation from the verity-application agent for the relationship created by this protocol
   *
   * @param context an instance of the Context object initialized to a verity-application agent
   * @param shortInvite enables short version of invite to be generated
   * @param goal goal code for providing the goal of the connection
   */
  async outOfBandInvitation (context, shortInvite = null, goal = null) {
    await this.sendMessage(context, await this.outOfBandInvitationMsgPacked(context, shortInvite, goal))
  }

  /**
   * Creates the control message without packaging and sending it.
   * @param context an instance of the Context object initialized to a verity-application agent
   * @param phoneNumber mobile phone number in international format (eg. +18011234567) which can be used for sending SMS invitations.
   * @param goal goal code for providing the goal of the connection
   * @return the constructed message (JSON object)
   *
   * @see #smsOutOfBandInvitation
   */
  async smsOutOfBandInvitationMsg (context, phoneNumber, goal = null) {
    var msg = this._getBaseMessage(this.msgNames.SMS_OUT_OF_BAND_INVITATION)
    msg['~for_relationship'] = this.forRelationship
    msg.phoneNumber = phoneNumber
    msg = this._addThread(msg)

    if (goal != null) {
      msg.goalCode = goal.code
      msg.goal = goal.goalName
    }
    return msg
  }

  /**
   * Creates and packages message without sending it.
   * @param context an instance of the Context object initialized to a verity-application agent
   * @param phoneNumber mobile phone number in international format (eg. +18011234567) which can be used for sending SMS invitations.
   * @param goal goal code for providing the goal of the connection
   * @return the byte array ready for transport
   *
   * @see #smsOutOfBandInvitation
   */
  async smsOutOfBandInvitationMsgPacked (context, phoneNumber, goal) {
    return this.getMessageBytes(context, await this.smsOutOfBandInvitationMsg(context, phoneNumber, goal))
  }

  /**
   * Ask for SMS aries out of band invitation from the verity-application agent for the relationship created by this protocol
   *
   * @param context an instance of the Context object initialized to a verity-application agent
   * @param phoneNumber mobile phone number in international format (eg. +18011234567) which can be used for sending SMS invitations.
   * @param goal goal code for providing the goal of the connection
   */
  async smsOutOfBandInvitation (context, phoneNumber, goal = null) {
    await this.sendMessage(context, await this.smsOutOfBandInvitationMsgPacked(context, phoneNumber, goal))
  }
}

module.exports = Relationship
